src/python/pants/backend/project_info/peek.py

```diff
@@ -0,0 +1,159 @@
+        )
+        register(
+            "--exclude-defaults",
+            type=bool,
+            default=False,
+            help="whether to exclude values matching default field values from normalized output",
+        )
+
+    @property
+    def output_type(self) -> OutputOptions:
+        """Get the output type from options.
+
+        Must be renamed here because `output` conflicts with `Outputting` class.
+        """
+        return cast(OutputOptions, self.options.output)
+
+    @property
+    def exclude_defaults(self) -> bool:
+        return cast(bool, self.options.exclude_defaults)
+
+
+class Peek(Goal):
+    subsystem_cls = PeekSubsystem
+
+
+def _render_raw(fcs: Iterable[FileContent]) -> str:
+    sorted_fcs = sorted(fcs, key=lambda fc: fc.path)
+    rendereds = map(_render_raw_build_file, sorted_fcs)
+    return os.linesep.join(rendereds)
+
+
+def _render_raw_build_file(fc: FileContent, encoding: str = "utf-8") -> str:
+    dashes = "-" * len(fc.path)
+    content = fc.content.decode(encoding)
+    parts = [dashes, fc.path, dashes, content]
+    if not content.endswith(os.linesep):
+        parts.append("")
+    return os.linesep.join(parts)
+
+
+def _render_json(ts: Iterable[Target], exclude_defaults: bool = False) -> str:
+    targets = [_target_to_dict(t, exclude_defaults) for t in ts]
+    data = dict(targets=targets)
+    return json.dumps(data, indent=2, cls=_PeekJsonEncoder)
+
+
+_nothing = object()
+
+
+def _target_to_dict(t: Target, exclude_defaults: bool = False, _nothing: object = _nothing) -> dict:
+    return {
+        "address": t.address.spec,
+        "target_type": t.alias,
+        **{
+            k.alias: v.value
+            for k, v in t.field_values.items()
+            # don't report default values in normalized output
+            if not (exclude_defaults and getattr(k, "default", _nothing) == v.value)
+        },
+    }
+
+
+class _PeekJsonEncoder(json.JSONEncoder):
+    """Allow us to serialize some commmonly-found types in BUILD files."""
+
+    safe_to_str_types = (Requirement,)
+
+    def default(self, o):
+        """Return a serializable object for o."""
+        if is_dataclass(o):
+            return asdict(o)
+        elif isinstance(o, collections.abc.Mapping):
+            return dict(o)
+        elif isinstance(o, collections.abc.Sequence):
+            return list(o)
+        elif isinstance(o, self.safe_to_str_types):
+            return str(o)
+        else:
+            try:
+                return super().default(o)
+            except TypeError:
+                # punt and just try str()
+                # TODO: can we find a better strategy here?
+                return str(o)
+
+
+@goal_rule
+async def peek(
+    console: Console,
+    subsys: PeekSubsystem,
+    addresses: Addresses,
+) -> Peek:
+    # TODO: better options for target selection (globs, transitive, etc)?
+    targets: Iterable[Target]
+    # TODO: handle target not found exceptions better here
```

Comment:
What do you mean? This should be handled already. I'd be very interested in suggestions for how to improve that error message though - input is always appreciated on error messages.